runners/flink/src/main/java/org/apache/beam/runners/flink/metrics/FlinkMetricContainer.java

```diff
@@ -89,127 +93,138 @@ public FlinkMetricContainer(RuntimeContext runtimeContext) {
     this.metricsAccumulator = (MetricsAccumulator) metricsAccumulator;
   }
 
-  public MetricsContainer getMetricsContainer(String stepName) {
+  @Nullable
+  public MetricsContainer getMetricsContainer(MetricLabels labels) {
     return metricsAccumulator != null
-        ? metricsAccumulator.getLocalValue().getContainer(stepName)
+        ? metricsAccumulator.getLocalValue().getContainer(labels)
         : null;
   }
 
+  @Nullable
+  public MetricsContainer ptransformContainer(String ptransform) {
+    return getMetricsContainer(MetricLabels.ptransform(ptransform));
+  }
+
   /**
    * Update this container with metrics from the passed {@link MonitoringInfo}s, and send updates
    * along to Flink's internal metrics framework.
    */
   public void updateMetrics(String stepName, List<MonitoringInfo> monitoringInfos) {
-    MetricsContainer metricsContainer = getMetricsContainer(stepName);
     monitoringInfos.forEach(
         monitoringInfo -> {
-          if (monitoringInfo.hasMetric()) {
```

Comment:
A lot of the Flink-related changes here have since been factored out to #7971, I'll try to get that merged, and in the meantime set up a view of this PR rebased on top of that.